<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to clonealign • clonealign</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to clonealign">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clonealign</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction_to_clonealign.html">Introduction to clonealign</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Introduction to clonealign</h1>
                        <h4 class="author">Kieran R Campbell</h4>
            <address class="author_afil">
      Department of Statistics, University of British Columbia<br><a class="author_email" href="mailto:#"></a><a href="mailto:kicampbell@bccrc.ca">kicampbell@bccrc.ca</a>
      </address>
                  
            <h4 class="date">October 2018</h4>
      
      
      <div class="hidden name"><code>introduction_to_clonealign.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p><code>clonealign</code> assigns cells measured using single-cell RNA-seq to their clones of origin using copy number data. This is especially useful when clones are inferred from shallow single-cell DNA-seq, in which case the copy number state of each clone is known, but the precise SNV structure is unknown.</p>
<p>To assign cells to clones, <code>clonealign</code> makes the assumption that</p>
<p><span class="math display">\[ \text{gene expression} \propto \text{number of gene copies} \]</span></p>
<p>This is demonstrated in the figure below.</p>
<p><img src="clonealign_figure.png" width="600" align="middle"></p>
<p>Mathematically we have an <span class="math inline">\(N \times G\)</span> matrix <span class="math inline">\(Y\)</span> of raw gene expression counts (from RNA-seq) for <span class="math inline">\(N\)</span> cells and <span class="math inline">\(G\)</span> genes, where <span class="math inline">\(y_{ng}\)</span> is the counts to gene <span class="math inline">\(g\)</span> in cell <span class="math inline">\(c\)</span>. We also have a <span class="math inline">\(G \times C\)</span> matrix <span class="math inline">\(\Lambda\)</span> of copy number variation for <span class="math inline">\(C\)</span> clones, where <span class="math inline">\(\lambda_{gc}\)</span> is the copy number of gene <span class="math inline">\(g\)</span> in clone <span class="math inline">\(c\)</span>. We introduce a clone-assigning categorical variable <span class="math inline">\(\pi_n\)</span> for each cell, where</p>
<p><span class="math display">\[z_n = c \text{ if cell $n$ on clone $c$} \]</span></p>
<p>then <code>clonealign</code> models the conditional expected counts in a gene and cell as</p>
<p><span class="math display">\[ E[y_{ng} | z_n=c] = \frac{\lambda_{g,c} f(\mu_g) e^{\psi_n \cdot w_g}}{ \sum_{g'}\lambda_{g',c} f(\mu_{g'}) e^{\psi_n \cdot w_{g'}}} s_n  \]</span></p>
<p>where <span class="math inline">\(s_n\)</span> is a cell-specific size factor, <span class="math inline">\(\mu_g\)</span> is the per-chromosome expression (normalized so that <span class="math inline">\(\mu_1 = 1\)</span> for model identifiability), <span class="math inline">\(f\)</span> is a function that maps copy number to a multiplicative factor of expression, and <span class="math inline">\(\psi\)</span> and <span class="math inline">\(w\)</span> are cell and gene specific random effects. The noise distribution is assumed to be negative-binomial. Inference is performed using reparametrization-gradient variational inference to calculate a posterior distribution of the clone assignments <span class="math inline">\(z_n\)</span> and of all other model parameters.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p><code>clonealign</code> is built upon <a href="https://www.tensorflow.org/">Google’s Tensorflow</a> using the <a href="https://tensorflow.rstudio.com/">Tensorflow R package provided by Rstudio</a>. To install <code>tensorflow</code>, run</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"tensorflow"</span>)
<span class="kw">library</span>(tensorflow)
<span class="kw"><a href="http://www.rdocumentation.org/packages/tensorflow/topics/install_tensorflow">install_tensorflow</a></span>()</code></pre></div>
<p>You can confirm the installation succeeded by running</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sess =<span class="st"> </span>tf<span class="op">$</span><span class="kw">Session</span>()
hello &lt;-<span class="st"> </span>tf<span class="op">$</span><span class="kw">constant</span>(<span class="st">'Hello, TensorFlow!'</span>)
sess<span class="op">$</span><span class="kw">run</span>(hello)</code></pre></div>
<p>For more details see the <a href="https://tensorflow.rstudio.com/tensorflow/articles/installation.html">Rstudio page on tensorflow installation</a>.</p>
<p><code>clonealign</code> can then be installed using the <code>devtools</code> package via</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"kieranrcampbell/clonealign"</span>)</code></pre></div>
</div>
<div id="basic-usage" class="section level1">
<h1 class="hasAnchor">
<a href="#basic-usage" class="anchor"></a>Basic usage</h1>
<div id="data-preparation" class="section level2">
<h2 class="hasAnchor">
<a href="#data-preparation" class="anchor"></a>Data preparation</h2>
<p>By default, <code>clonealign</code> requires two inputs:</p>
<ul>
<li>Gene expression data as raw counts. This can be in the form of a <code>SingleCellExperiment</code>, <code>SummarizedExperiment</code> or <strong>cell by gene</strong> <code>matrix</code>
</li>
<li>Copy number profiles for each clone and gene (where the genes <strong>must</strong> be the same as those measured in the expression data). This can be in the form of a <code>data.frame</code>, <code>DataFrame</code> or <code>matrix</code>
</li>
</ul>
<p>Bundled with the package is an example <code>SingleCellExperiment</code> for 100 genes and 200 cells:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(clonealign)
<span class="kw">data</span>(example_sce)
example_sce
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 100 200 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(100): gene_1 gene_2 ... gene_99 gene_100</span>
<span class="co">#&gt; rowData names(3): A B C</span>
<span class="co">#&gt; colnames(200): cell_1 cell_2 ... cell_199 cell_200</span>
<span class="co">#&gt; colData names(0):</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; spikeNames(0):</span></code></pre></div>
<p>This has raw integer counts in the <code>assays</code> slot as required for input to <code>clonealign</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assay</span>(example_sce, <span class="st">"counts"</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
<span class="co">#&gt;        cell_1 cell_2 cell_3 cell_4 cell_5</span>
<span class="co">#&gt; gene_1      0      0      0      0      0</span>
<span class="co">#&gt; gene_2      0      0      0      0      0</span>
<span class="co">#&gt; gene_3      1      1      1      0      0</span>
<span class="co">#&gt; gene_4      0      0      0      0      0</span>
<span class="co">#&gt; gene_5      1      0      0      1      0</span></code></pre></div>
<p>The CNV data is stored in the <code>rowData</code> of the <code>SingleCellExperiment</code> for 3 clones (A, B, and C) and crucially the same genes as the expression data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cnv_data &lt;-<span class="st"> </span><span class="kw">rowData</span>(example_sce)[, <span class="kw">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>)]
<span class="kw">stopifnot</span>(<span class="kw">nrow</span>(cnv_data) <span class="op">==</span><span class="st"> </span><span class="kw">nrow</span>(example_sce)) <span class="co"># Make sure genes match up</span>
<span class="kw">head</span>(cnv_data)
<span class="co">#&gt; DataFrame with 6 rows and 3 columns</span>
<span class="co">#&gt;           A         B         C</span>
<span class="co">#&gt;   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">#&gt; 1         1         2         2</span>
<span class="co">#&gt; 2         2         1         1</span>
<span class="co">#&gt; 3         3         2         2</span>
<span class="co">#&gt; 4         3         2         2</span>
<span class="co">#&gt; 5         2         2         3</span>
<span class="co">#&gt; 6         2         1         1</span></code></pre></div>
</div>
<div id="model-fitting" class="section level2">
<h2 class="hasAnchor">
<a href="#model-fitting" class="anchor"></a>Model fitting</h2>
<p>The model is fitted with a basic call to <code>clonealign</code>, which prints the ELBO for each iteration (this can be turned off by setting <code>verbose = FALSE</code>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cal &lt;-<span class="st"> </span><span class="kw"><a href="../reference/clonealign.html">clonealign</a></span>(example_sce, cnv_data)
<span class="co">#&gt; Removing 0 genes with low counts</span>
<span class="co">#&gt; Creating Tensorflow graph...</span>
<span class="co">#&gt; clonealign inference complete</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(cal)
<span class="co">#&gt; A clonealign_fit for 200 cells, 100 genes, and 3 clones</span>
<span class="co">#&gt; To access clone assignments, call x$clone</span>
<span class="co">#&gt; To access ML parameter estimates, call x$ml_params</span></code></pre></div>
<p>We can plot the ELBO to ensure convergence:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(<span class="kw">seq_along</span>(cal<span class="op">$</span>elbo), cal<span class="op">$</span>elbo, <span class="dt">geom =</span> <span class="kw">c</span>(<span class="st">"point"</span>, <span class="st">"line"</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">"Iteration"</span>, <span class="dt">y =</span> <span class="st">"ELBO"</span>)</code></pre></div>
<p><img src="introduction_to_clonealign_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>The maximum likelihood estimates of the clone assignments can be access through the <code>clone</code> slot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clones &lt;-<span class="st"> </span>cal<span class="op">$</span>clone
<span class="kw">table</span>(clones)
<span class="co">#&gt; clones</span>
<span class="co">#&gt;   A   B   C </span>
<span class="co">#&gt; 184   7   9</span></code></pre></div>
<p>This can easily be added to the <code>SingleCellExperiment</code> for visualization with <code>scater</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">example_sce<span class="op">$</span>clone &lt;-<span class="st"> </span>clones
example_sce &lt;-<span class="st"> </span><span class="kw">normalise</span>(example_sce)
<span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span>
<span class="kw">plotPCA</span>(example_sce, <span class="dt">ncomponents =</span> <span class="dv">3</span>, <span class="dt">colour_by =</span> <span class="st">"clone"</span>)</code></pre></div>
<p><img src="introduction_to_clonealign_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>The clone assignments in <code>clones</code> can then be used for the desired downstream analysis, such as differential expression or SNV analysis.</p>
</div>
<div id="plotting-results" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-results" class="anchor"></a>Plotting results</h2>
<p>The <code>plot_clonealign</code> function can be used to check the sanity of the fitted clones by ensuring that gene expression does correlate to the inferred copy number. For this we require the input <code>SingleCellExperiment</code>, copy number matrix, and clone assignments. Note that the <code>SingleCellExperiment</code> requires columns in <code>rowData</code> corresponding to the chromosome, start and end position of each feature (gene). These can conveniently be gathered using the <code>getBMFeatureAnnos</code> function in <code>scater</code>, e.g.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sce &lt;-<span class="st"> </span><span class="kw">getBMFeatureAnnos</span>(sce, <span class="dt">filters =</span> <span class="st">"ensembl_gene_id"</span>,
                         <span class="dt">attributes =</span> <span class="kw">c</span>(<span class="st">"ensembl_gene_id"</span>, <span class="st">"start_position"</span>, <span class="st">"end_position"</span>),
                         <span class="dt">feature_symbol =</span> <span class="st">"hgnc_symbol"</span>,
                         <span class="dt">feature_id =</span> <span class="st">"ensembl_gene_id"</span>,
                         <span class="dt">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>)</code></pre></div>
<p>For now we’ll set these to made-up values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene_position &lt;-<span class="st"> </span><span class="kw">as_data_frame</span>(cnv_data) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">gene =</span> <span class="kw">seq_len</span>(<span class="kw">nrow</span>(cnv_data))) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">arrange</span>(A, B, C) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">position =</span> <span class="kw">seq_len</span>(<span class="kw">nrow</span>(cnv_data))) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">arrange</span>(gene) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>.<span class="op">$</span>position
<span class="co">#&gt; Warning in as.data.frame(x, row.names = NULL, optional = optional, ...):</span>
<span class="co">#&gt; Arguments in '...' ignored</span>

<span class="kw">rowData</span>(example_sce)<span class="op">$</span>chromosome &lt;-<span class="st"> "1"</span>
<span class="kw">rowData</span>(example_sce)<span class="op">$</span>start_pos &lt;-<span class="st"> </span>gene_position
<span class="kw">rowData</span>(example_sce)<span class="op">$</span>end_pos &lt;-<span class="st"> </span>gene_position</code></pre></div>
<p>We can then plot the expression estimates using the <code>plot_clonealign</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_clonealign.html">plot_clonealign</a></span>(example_sce, cal<span class="op">$</span>clone, cnv_data,
                <span class="dt">chromosome =</span> <span class="st">"1"</span>,
                <span class="dt">chr_str =</span> <span class="st">"chromosome"</span>,
                <span class="dt">start_str =</span> <span class="st">"start_pos"</span>,
                <span class="dt">end_str =</span> <span class="st">"end_pos"</span>)
<span class="co">#&gt; Joining, by = "ensembl_gene_id"</span>
<span class="co">#&gt; Joining, by = c("clone", "state")</span>
<span class="co">#&gt; Joining, by = c("state", "clone")</span></code></pre></div>
<p><img src="introduction_to_clonealign_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>where the <code>*_str</code> identifies the columns of <code>rowData(example_sce)</code> to look for the chromosome names and feature start and end positions.</p>
</div>
<div id="evaluating-the-clone-assignment" class="section level2">
<h2 class="hasAnchor">
<a href="#evaluating-the-clone-assignment" class="anchor"></a>Evaluating the clone assignment</h2>
<p><code>clonealign</code> will assign single-cell RNA-seq to clones no matter how good the fit and no matter whether the clones actually exist in the expression data. As a consequence, it is important to evaluate the quality of fit which in practice we do by fitting the clones on a subset of genes and looking at how well the predicted expression profiles fits the observed expression data of the “held-out” genes.</p>
<p>We perform his out-of-sample evaluation using the <code>evaluate_clonealign</code> function, which takes in the clonealign fit on the full dataset (along with the dataset) and both prints various metrics and returns a list containing various computed objects:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ec &lt;-<span class="st"> </span><span class="kw"><a href="../reference/evaluate_clonealign.html">evaluate_clonealign</a></span>(example_sce, cnv_data, cal)
<span class="co">#&gt; On the full dataset, the observed MSE was on average 1.01506017179589 times smaller than under a null model and smaller 100% of the time</span>
<span class="co">#&gt; Fitting reduced clonealign model...</span>
<span class="co">#&gt; Agreement between original (rows) and reduced (columns):   </span>
<span class="co">#&gt;       A   B   C</span>
<span class="co">#&gt;   A 105   0   0</span>
<span class="co">#&gt;   B  78   7   4</span>
<span class="co">#&gt;   C   1   0   5</span>
<span class="co">#&gt; On the held-out dataset, the observed MSE was on average 1.02471611655892 times smaller than under a null model and smaller 100% of the time</span></code></pre></div>
<p>The function call prints several important metrics:</p>
<ol style="list-style-type: decimal">
<li>The mean square error (MSE) on the full dataset given the clone assignments, compared to the average MSE under a null distribution found by permuting the clone assignments. If the observed MSE is greater than the null MSE (either on average or a large percent of the time) then the fit is very poor and should be discarded</li>
<li>The agreement between the clone assignments on the full geneset and under the null geneset. If these are very different then the algorithm is sensitive to the input genes and caution is advised</li>
<li>The MSE of held-out (out-of-sample) genes using a clonealign fit on an orthogonal test set of genes. Again, if the observed MSE is greater than the null MSE (or approximately the same) then the fit is unreliable and should be discarded.</li>
</ol>
<p>These metrics are stored in a list returned by the function call. For details, see the help page at <code><a href="../reference/evaluate_clonealign.html">?evaluate_clonealign</a></code>.</p>
</div>
</div>
<div id="advanced-options" class="section level1">
<h1 class="hasAnchor">
<a href="#advanced-options" class="anchor"></a>Advanced options</h1>
<div id="controlling-variational-inference" class="section level2">
<h2 class="hasAnchor">
<a href="#controlling-variational-inference" class="anchor"></a>Controlling Variational Inference</h2>
<p>Inference is performed using reparametrization gradient variational inference which uses the evidence lower bound (ELBO) to monitor convergence. This is controlled using the <code>rel_tol</code> parameter. When the difference</p>
<p><span class="math display">\[ \Delta ELBO = \frac{ELBO_{\text{new}} - ELBO_{\text{old}}}{|ELBO_{\text{old}}|} \]</span> falls below <code>rel_tol</code>, the optimization algorithm is considered converged. The maximum number of iterations to acheive this is set using the <code>max_iter</code> parameter.</p>
<p>Inference is performed using Adam (<span class="citation">Kingma and Ba (2014)</span>) which is controlled by three parameters:</p>
<ul>
<li>
<code>learning_rate</code> the learning rate</li>
<li>
<code>rel_tol</code> the relative difference in the ELBO below which the optimization will be considered converged (see above)</li>
<li>
<code>max_iter</code> the maximum number of Adam iterations to perform (see above)</li>
</ul>
</div>
<div id="accessing-posterior-maximum-likelihood-parameter-estimates" class="section level2">
<h2 class="hasAnchor">
<a href="#accessing-posterior-maximum-likelihood-parameter-estimates" class="anchor"></a>Accessing posterior / maximum-likelihood parameter estimates</h2>
<p>The object returned by a call to <code>clonealign</code> contains a <code>clone</code> slot for the maximum likelihood (ML) clone assignment for each cell. The ML estimates of the other parameters can be found in the <code>cal$ml_params</code> slot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(cal<span class="op">$</span>ml_params)
<span class="co">#&gt; [1] "clone_probs" "mu"          "s"           "alpha"       "a"          </span>
<span class="co">#&gt; [6] "b"           "psi"         "W"           "chi"</span></code></pre></div>
<p>The slot <code>clone_probs</code> gives the probability that each cell is assigned to each clone:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(cal<span class="op">$</span>ml_params<span class="op">$</span>clone_probs)
<span class="co">#&gt;               A           B           C</span>
<span class="co">#&gt; [1,] 0.97769814 0.020225340 0.002076519</span>
<span class="co">#&gt; [2,] 0.02084794 0.480126307 0.499025749</span>
<span class="co">#&gt; [3,] 0.98862077 0.007752682 0.003626547</span>
<span class="co">#&gt; [4,] 0.95802716 0.039739091 0.002233748</span>
<span class="co">#&gt; [5,] 0.05501944 0.253582424 0.691398136</span>
<span class="co">#&gt; [6,] 0.95206569 0.044654717 0.003279597</span></code></pre></div>
<p>while <code>mu</code> and <code>phi</code> give the maximum likelihood estimates of the <span class="math inline">\(\mu\)</span> and <span class="math inline">\(\phi\)</span> parameters from the model.</p>
</div>
</div>
<div id="technical" class="section level1">
<h1 class="hasAnchor">
<a href="#technical" class="anchor"></a>Technical</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()
<span class="co">#&gt; R version 3.5.0 (2018-04-23)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Debian GNU/Linux 9 (stretch)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS: /usr/lib/openblas-base/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/libopenblasp-r0.2.19.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] bindrcpp_0.2.2              clonealign_0.99.0          </span>
<span class="co">#&gt;  [3] dplyr_0.7.5                 scater_1.8.0               </span>
<span class="co">#&gt;  [5] ggplot2_3.0.0               SingleCellExperiment_1.2.0 </span>
<span class="co">#&gt;  [7] SummarizedExperiment_1.10.1 DelayedArray_0.6.1         </span>
<span class="co">#&gt;  [9] BiocParallel_1.14.1         matrixStats_0.53.1         </span>
<span class="co">#&gt; [11] Biobase_2.40.0              GenomicRanges_1.32.3       </span>
<span class="co">#&gt; [13] GenomeInfoDb_1.16.0         IRanges_2.14.10            </span>
<span class="co">#&gt; [15] S4Vectors_0.18.3            BiocGenerics_0.26.0        </span>
<span class="co">#&gt; [17] BiocStyle_2.8.2            </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-6             fs_1.2.2                </span>
<span class="co">#&gt;  [3] RColorBrewer_1.1-2       progress_1.2.0          </span>
<span class="co">#&gt;  [5] rprojroot_1.3-2          tools_3.5.0             </span>
<span class="co">#&gt;  [7] backports_1.1.2          R6_2.2.2                </span>
<span class="co">#&gt;  [9] vipor_0.4.5              lazyeval_0.2.1          </span>
<span class="co">#&gt; [11] colorspace_1.3-2         withr_2.1.2             </span>
<span class="co">#&gt; [13] prettyunits_1.0.2        tidyselect_0.2.4        </span>
<span class="co">#&gt; [15] gridExtra_2.3            compiler_3.5.0          </span>
<span class="co">#&gt; [17] xml2_1.2.0               desc_1.2.0              </span>
<span class="co">#&gt; [19] labeling_0.3             bookdown_0.7            </span>
<span class="co">#&gt; [21] scales_1.0.0             tfruns_1.4              </span>
<span class="co">#&gt; [23] pkgdown_1.0.0            commonmark_1.5          </span>
<span class="co">#&gt; [25] stringr_1.3.1            digest_0.6.16           </span>
<span class="co">#&gt; [27] rmarkdown_1.10           XVector_0.20.0          </span>
<span class="co">#&gt; [29] base64enc_0.1-3          pkgconfig_2.0.1         </span>
<span class="co">#&gt; [31] htmltools_0.3.6          limma_3.36.2            </span>
<span class="co">#&gt; [33] rlang_0.2.2              rstudioapi_0.7          </span>
<span class="co">#&gt; [35] shiny_1.1.0              DelayedMatrixStats_1.2.0</span>
<span class="co">#&gt; [37] bindr_0.1.1              jsonlite_1.5            </span>
<span class="co">#&gt; [39] tensorflow_1.5           RCurl_1.95-4.10         </span>
<span class="co">#&gt; [41] magrittr_1.5             GenomeInfoDbData_1.1.0  </span>
<span class="co">#&gt; [43] Matrix_1.2-14            Rcpp_0.12.19            </span>
<span class="co">#&gt; [45] ggbeeswarm_0.6.0         munsell_0.5.0           </span>
<span class="co">#&gt; [47] Rhdf5lib_1.2.1           reticulate_1.10         </span>
<span class="co">#&gt; [49] viridis_0.5.1            whisker_0.3-2           </span>
<span class="co">#&gt; [51] stringi_1.2.2            yaml_2.1.19             </span>
<span class="co">#&gt; [53] edgeR_3.22.3             MASS_7.3-50             </span>
<span class="co">#&gt; [55] zlibbioc_1.26.0          rhdf5_2.24.0            </span>
<span class="co">#&gt; [57] plyr_1.8.4               grid_3.5.0              </span>
<span class="co">#&gt; [59] promises_1.0.1           shinydashboard_0.7.0    </span>
<span class="co">#&gt; [61] crayon_1.3.4             lattice_0.20-35         </span>
<span class="co">#&gt; [63] cowplot_0.9.2            hms_0.4.2               </span>
<span class="co">#&gt; [65] locfit_1.5-9.1           knitr_1.20              </span>
<span class="co">#&gt; [67] pillar_1.2.3             rjson_0.2.20            </span>
<span class="co">#&gt; [69] reshape2_1.4.3           glue_1.2.0              </span>
<span class="co">#&gt; [71] evaluate_0.10.1          data.table_1.11.4       </span>
<span class="co">#&gt; [73] httpuv_1.4.3             tidyr_0.8.1             </span>
<span class="co">#&gt; [75] gtable_0.2.0             purrr_0.2.5             </span>
<span class="co">#&gt; [77] assertthat_0.2.0         xfun_0.2                </span>
<span class="co">#&gt; [79] mime_0.5                 xtable_1.8-2            </span>
<span class="co">#&gt; [81] roxygen2_6.0.1           later_0.7.3             </span>
<span class="co">#&gt; [83] viridisLite_0.3.0        tibble_1.4.2            </span>
<span class="co">#&gt; [85] beeswarm_0.2.3           memoise_1.1.0           </span>
<span class="co">#&gt; [87] tximport_1.8.0</span></code></pre></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-kingma2014adam">
<p>Kingma, Diederik P, and Jimmy Ba. 2014. “Adam: A Method for Stochastic Optimization.” <em>arXiv Preprint arXiv:1412.6980</em>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#installation">Installation</a></li>
      <li>
<a href="#basic-usage">Basic usage</a><ul class="nav nav-pills nav-stacked">
<li><a href="#data-preparation">Data preparation</a></li>
      <li><a href="#model-fitting">Model fitting</a></li>
      <li><a href="#plotting-results">Plotting results</a></li>
      <li><a href="#evaluating-the-clone-assignment">Evaluating the clone assignment</a></li>
      </ul>
</li>
      <li>
<a href="#advanced-options">Advanced options</a><ul class="nav nav-pills nav-stacked">
<li><a href="#controlling-variational-inference">Controlling Variational Inference</a></li>
      <li><a href="#accessing-posterior-maximum-likelihood-parameter-estimates">Accessing posterior / maximum-likelihood parameter estimates</a></li>
      </ul>
</li>
      <li><a href="#technical">Technical</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Kieran Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
