<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Assign scRNA-seq to clone-of-origin — clonealign • clonealign</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Assign scRNA-seq to clone-of-origin — clonealign" />

<meta property="og:description" content="clonealign assigns single cells (measured with RNA-seq) to their clones of origin, where
the clones have been inferred from ultra-shallow scDNA-seq and collated into copy number profiles." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clonealign</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction_to_clonealign.html">Introduction to clonealign</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Assign scRNA-seq to clone-of-origin</h1>
    
    <div class="hidden name"><code>clonealign.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p><code>clonealign</code> assigns single cells (measured with RNA-seq) to their clones of origin, where
the clones have been inferred from ultra-shallow scDNA-seq and collated into copy number profiles.</p>
    
    </div>

    <pre class="usage"><span class='fu'>clonealign</span>(<span class='no'>gene_expression_data</span>, <span class='no'>copy_number_data</span>, <span class='kw'>max_iter_em</span> <span class='kw'>=</span> <span class='fl'>100</span>,
  <span class='kw'>max_iter_adam</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>rel_tol_em</span> <span class='kw'>=</span> <span class='fl'>1e-06</span>, <span class='kw'>rel_tol_adam</span> <span class='kw'>=</span> <span class='fl'>1e-06</span>,
  <span class='kw'>gene_filter_threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>learning_rate</span> <span class='kw'>=</span> <span class='fl'>0.1</span>, <span class='kw'>fix_alpha</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>clone_specific_phi</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>fix_s</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sigma_hyper</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>dtype</span> <span class='kw'>=</span> <span class='st'>"float64"</span>, <span class='kw'>saturate</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gene_expression_data</th>
      <td><p>A matrix of gene counts or a
<code>SingleCellExperiment</code>. This should contain raw counts. See <code>details</code>.</p></td>
    </tr>
    <tr>
      <th>copy_number_data</th>
      <td><p>A matrix or data frame of copy number calls for each clone.
See <code>details</code>.</p></td>
    </tr>
    <tr>
      <th>max_iter_em</th>
      <td><p>Maximum number of EM iterations to perform</p></td>
    </tr>
    <tr>
      <th>max_iter_adam</th>
      <td><p>The maximum number of Adam iterations to perform in each M-step</p></td>
    </tr>
    <tr>
      <th>rel_tol_em</th>
      <td><p>Relative tolerance (change in log-likelihood per EM iteration in percent) below which the
EM algorithm is considered converged</p></td>
    </tr>
    <tr>
      <th>rel_tol_adam</th>
      <td><p>The relative tolerance for each Adam update in the M-step below which the
maximization will be considered converged</p></td>
    </tr>
    <tr>
      <th>gene_filter_threshold</th>
      <td><p>Genes with total counts below or equal to this threshold will
be filtered out (removes genes with no counts by default)</p></td>
    </tr>
    <tr>
      <th>learning_rate</th>
      <td><p>The learning rate to be passed to the Adam optimizer</p></td>
    </tr>
    <tr>
      <th>fix_alpha</th>
      <td><p>Should the underlying priors for clone frequencies be fixed? Default TRUE
(values are inferred from the data)</p></td>
    </tr>
    <tr>
      <th>clone_specific_phi</th>
      <td><p>Should each clone have distinct over-dispersion parameters? Default TRUE</p></td>
    </tr>
    <tr>
      <th>fix_s</th>
      <td><p>Should the size factors be fixed? If <code>NULL</code> they are jointly inferred from the data, otherwise
a vector corresponding to the number of cells should be specified.</p></td>
    </tr>
    <tr>
      <th>sigma_hyper</th>
      <td><p>The hyperparameter governing shrinkage of clone-specific dispersions</p></td>
    </tr>
    <tr>
      <th>dtype</th>
      <td><p>The dtype for tensorflow useage, either "float32" or "float64"</p></td>
    </tr>
    <tr>
      <th>saturate</th>
      <td><p>Should the CNV-expression relationship saturate above copy number = 4? Default TRUE</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Should warnings and EM convergence information be printed? Default TRUE</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of class <code>clonealign_fit</code>. The maximum likelihood estimates of the
clone assignment paramters are in the <code>clone</code> slot. Maximum likelihood estimates of
all model parameters are in the <code>ml_params</code> slot.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><strong>Input format</strong></p>
<p><code>gene_expression_data</code> must either be a <code>SingleCellExperiment</code> or <code>SummarizedExperiment</code>
with a <code>counts</code> assay
representing raw gene expression counts, or a cell by gene matrix of raw counts.</p>
<p><code>copy_number_data</code> must either be a <code>matrix</code>, <code>data.frame</code> or <code>DataFrame</code> with a
row for each gene in <code>gene_expression_data</code> and a column for each of the clones.
If <code>colnames(copy_number_data)</code> is not <code>NULL</code> then these names will be used for each of
the clones in the final output.</p>
<p><strong>Recommended parameter settings</strong></p>
<p>As with any probabilistic model there are many parameters to set. Through comprehensive simulations regarding
the robustness of the model to mis-specification (ie what's the minimum proportion of genes for which the
CNV-expression relationship can be true and our inferences still valid) we have come up with the following
guidelines for parameter settings, reflected in the default values:</p>
<ul>
<li><p>Number of ADAM iterations - if set to 1 we essentially perform gradient descent on the marginal log-likelihood
which empircally appears to have the best performance</p></li>
<li><p>Dispersions should be clone-specific with weak shrinkage (<code>sigma</code> = 1 appears best)</p></li>
<li><p>The generating probabilities should be fixed to be a priori equal (this corresponds to setting <code>alpha = TRUE</code>)</p></li>
<li><p>The cell size factors are best fixed in advanced by multiplying the total counts of whatever genes are passed
to clonealign by the edgeR (TMM) normalization factors</p></li>
</ul>
    <p><strong>Controlling the EM algorithm</strong></p>
<p>Inference is performed using the EM algorithm (<a href='https://en.wikipedia.org/wiki/Expectation-maximization_algorithm'>https://en.wikipedia.org/wiki/Expectation-maximization_algorithm</a>)
which uses the log marginal likelihood to monitor convergence. This is controlled using the
<code>rel_tol_em</code> parameter. When the difference between the new and old log marginal likelihoods normalized
by the absolute value of the old falls below <code>rel_tol_em</code>, the EM algorithm is considered converged.
The maximum number of iterations to acheive this is set using the <code>max_iter_em</code> parameter.</p>
<p>In each M-step, maximization is performed using Adam, with learning rate given by <code>learning_rate</code>. Each M-step
is considered converged with the value of the Q function falls below <code>rel_tol_adam</code> with a maximum number of iterations
<code>max_iter_adam</code>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>library</span>(<span class='no'>SingleCellExperiment</span>)
<span class='fu'>data</span>(<span class='no'>example_sce</span>)
<span class='no'>copy_number_data</span> <span class='kw'>&lt;-</span> <span class='fu'>rowData</span>(<span class='no'>example_sce</span>)[,<span class='fu'>c</span>(<span class='st'>"A"</span>, <span class='st'>"B"</span>, <span class='st'>"C"</span>)]
<span class='no'>cal</span> <span class='kw'>&lt;-</span> <span class='fu'>clonealign</span>(<span class='no'>example_sce</span>, <span class='no'>copy_number_data</span>)</div><div class='output co'>#&gt; <span class='message'>Removing 0 genes with low counts</span></div><div class='output co'>#&gt; <span class='message'>Creating Tensorflow graph...</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Maximim number of ADAM iterations reached reached</span></div><div class='output co'>#&gt; <span class='message'>clonealign inference complete</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>cal</span>)</div><div class='output co'>#&gt; A clonealign_fit for 200 cells, 100 genes, and 3 clones
#&gt; To access clone assignments, call x$clone
#&gt; To access ML parameter estimates, call x$ml_params</div><div class='input'><span class='no'>clones</span> <span class='kw'>&lt;-</span> <span class='no'>cal</span>$<span class='no'>clone</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Kieran Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

