<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot gene expression and copy number — plot_clonealign • clonealign</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Plot gene expression and copy number — plot_clonealign" />

<meta property="og:description" content="Plot gene expression and copy number as a function of genomic coordinate." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clonealign</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction_to_clonealign.html">Introduction to clonealign</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot gene expression and copy number</h1>
    
    <div class="hidden name"><code>plot_clonealign.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Plot gene expression and copy number as a function of genomic coordinate.</p>
    
    </div>

    <pre class="usage"><span class='fu'>plot_clonealign</span>(<span class='no'>sce</span>, <span class='no'>clones</span>, <span class='no'>cnv_data</span>, <span class='kw'>chromosome</span> <span class='kw'>=</span> <span class='st'>"1"</span>, <span class='kw'>chr_str</span> <span class='kw'>=</span> <span class='st'>"chr"</span>,
  <span class='kw'>start_str</span> <span class='kw'>=</span> <span class='st'>"start_position"</span>, <span class='kw'>end_str</span> <span class='kw'>=</span> <span class='st'>"end_position"</span>,
  <span class='kw'>jitter_cnv</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>ggplot_palette</span> <span class='kw'>=</span> <span class='st'>"Set1"</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sce</th>
      <td><p>A <code>SingleCellExperiment</code>. Note this must have the fields
<code>chr_str</code>, <code>start_str</code>, and <code>end_str</code> in <code>rowData</code>. See details.</p></td>
    </tr>
    <tr>
      <th>clones</th>
      <td><p>The clone assignments of each cell - must be of the same set as the colnames of <code>L</code></p></td>
    </tr>
    <tr>
      <th>cnv_data</th>
      <td><p>The gene by clone copy number matrix used as input to <code>clonealign_fit</code>. The
genes (rows) of this must match to the genes in <code>sce</code></p></td>
    </tr>
    <tr>
      <th>chromosome</th>
      <td><p>The chromosome to plot</p></td>
    </tr>
    <tr>
      <th>chr_str</th>
      <td><p>The column of <code>rowData(sce)</code> that refers to the chromosome of each feature</p></td>
    </tr>
    <tr>
      <th>start_str</th>
      <td><p>The column of <code>rowData(sce)</code> that refers to the start position of each feature</p></td>
    </tr>
    <tr>
      <th>end_str</th>
      <td><p>The column of <code>rowData(sce)</code> that refers to the end position of each feature</p></td>
    </tr>
    <tr>
      <th>jitter_cnv</th>
      <td><p>Logical - if true random noise is added to the copy number profiles to help
display all clones in case two lie on top of each other</p></td>
    </tr>
    <tr>
      <th>ggplot_palette</th>
      <td><p>The palette to be passed to <code>scale_colour_brewer</code></p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A ggplot2 object displaying a CNV and gene expression track for each (inferred) clone</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function requires the chromosome, start, and end positions of each features to plot.
These are encoded as the <code>chr_str</code>, <code>start_str</code>, and <code>end_str</code> features
in <code>rowData(sce)</code> respectively. If we have ensembl ids or similar we can get add the
required fields using the <code>getBMFeatureAnnos</code> function in the <code>scater</code> package - <br />
<code>
sce &lt;- getBMFeatureAnnos(sce, filters = "ensembl_gene_id",
                         attributes = c("chromosome_name", "start_position", "end_position"),
                         feature_symbol = "hgnc_symbol",
                         feature_id = "ensembl_gene_id",
                         dataset = "hsapiens_gene_ensembl")
</code> <br />
Then we would call <code>plot_clonealign</code> using <code>chr_str == "chromosome_name"</code>, <code>start_str == "start_position"</code>,
and <code>end_str == "end_position"</code>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>library</span>(<span class='no'>SingleCellExperiment</span>)
<span class='fu'>library</span>(<span class='no'>scater</span>)
<span class='fu'>library</span>(<span class='no'>dplyr</span>)
<span class='fu'>data</span>(<span class='no'>example_clonealign_fit</span>)
<span class='no'>cnv_data</span> <span class='kw'>&lt;-</span> <span class='fu'>rowData</span>(<span class='no'>example_sce</span>)[,<span class='fu'>c</span>(<span class='st'>"A"</span>, <span class='st'>"B"</span>, <span class='st'>"C"</span>)]
<span class='no'>gene_position</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='http://dplyr.tidyverse.org/reference/reexports.html'>as_data_frame</a></span>(<span class='no'>cnv_data</span>) <span class='kw'>%&gt;%</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/scater/topics/mutate'>mutate</a></span>(<span class='kw'>gene</span> <span class='kw'>=</span> <span class='fu'>seq_len</span>(<span class='fu'>nrow</span>(<span class='no'>cnv_data</span>))) <span class='kw'>%&gt;%</span>
  <span class='fu'><a href='http://www.rdocumentation.org/packages/scater/topics/arrange'>arrange</a></span>(<span class='no'>A</span>, <span class='no'>B</span>, <span class='no'>C</span>) <span class='kw'>%&gt;%</span>
 <span class='fu'><a href='http://www.rdocumentation.org/packages/scater/topics/mutate'>mutate</a></span>(<span class='kw'>position</span> <span class='kw'>=</span> <span class='fu'>seq_len</span>(<span class='fu'>nrow</span>(<span class='no'>cnv_data</span>))) <span class='kw'>%&gt;%</span>
  <span class='fu'><a href='http://www.rdocumentation.org/packages/scater/topics/arrange'>arrange</a></span>(<span class='no'>gene</span>) <span class='kw'>%&gt;%</span>
  <span class='no'>.</span>$<span class='no'>position</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Arguments in '...' ignored</span></div><div class='input'>
<span class='fu'>rowData</span>(<span class='no'>example_sce</span>)$<span class='no'>chromosome</span> <span class='kw'>&lt;-</span> <span class='st'>"1"</span>
<span class='fu'>rowData</span>(<span class='no'>example_sce</span>)$<span class='no'>start_pos</span> <span class='kw'>&lt;-</span> <span class='no'>gene_position</span>
<span class='fu'>rowData</span>(<span class='no'>example_sce</span>)$<span class='no'>end_pos</span> <span class='kw'>&lt;-</span> <span class='no'>gene_position</span>
<span class='no'>example_sce</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/scater/topics/normalize'>normalise</a></span>(<span class='no'>example_sce</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: using library sizes as size factors</span></div><div class='input'><span class='fu'>plot_clonealign</span>(<span class='no'>example_sce</span>, <span class='no'>example_clonealign_fit</span>$<span class='no'>clone</span>, <span class='no'>cnv_data</span>,
<span class='kw'>chromosome</span> <span class='kw'>=</span> <span class='st'>"1"</span>,
<span class='kw'>chr_str</span> <span class='kw'>=</span> <span class='st'>"chromosome"</span>,
<span class='kw'>start_str</span> <span class='kw'>=</span> <span class='st'>"start_pos"</span>,
<span class='kw'>end_str</span> <span class='kw'>=</span> <span class='st'>"end_pos"</span>)</div><div class='output co'>#&gt; <span class='message'>Joining, by = "ensembl_gene_id"</span></div><div class='output co'>#&gt; <span class='message'>Joining, by = c("clone", "state")</span></div><div class='output co'>#&gt; <span class='message'>Joining, by = c("state", "clone")</span></div><div class='img'><img src='plot_clonealign-1.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Kieran Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

